### PR TITLE
fix(types): permissive props for un-introspectable targets, plus as-target autocomplete while typing

### LINKED ISSUE #5756: `styled(MantineButton)` rejects all props in 6.4.2 (regression from 6.1.x)
In 6.1.x, wrapping a component whose `React.ComponentPropsWithRef<typeof T>` resolves to `{}` produced a styled component that was permissive enough to accept arbitrary props at JSX call sites. In 6.4.2 the same code rejects every prop, including `children`.

This breaks every `styled(X)` where `X` is a Mantine v7 polymorphic-factory component (`Button`, `Card`, `Menu.Item`, `Badge`, `SegmentedControl`, …). TypeScript can't introspect Mantine's callable-generic intersection, so `React.ComponentPropsWithRef<typeof X>` resolves to `{}` in any sc version. 6.1.x stayed permissive on that empty BaseProps; 6.4.2 does not.

#### Simple Example

```tsx
import * as React from 'react';
import styled from 'styled-components';
import { Button as MantineButton } from '@mantine/core';

const Styled = styled(MantineButton)`color: red;`;

// 6.1.x: OK
// 6.4.2: TS2769 — `variant` and `children` do not exist on
//   `FastOmit<FastOmit<{}, never> & Partial<Pick<{}, never>>, "as" | "forwardedAs">
//    & { as?: WebTarget; forwardedAs?: WebTarget }`
const usage = <Styled variant="filled">hi</Styled>;
```

### PR DESCRIPTION
**Pull request by quantizor**
This PR lands two related TypeScript fixes for polymorphic styled components.

## 1. Permissive props for un-introspectable targets (#5756)

Wrapping a component whose props can't be statically read produced a styled component that rejected every prop at the JSX call site, including `children`. This hit every `styled(X)` where `X` is a Mantine v7 polymorphic-factory component (`Button`, `Card`, `Menu.Item`, `Badge`, `SegmentedControl`, and friends), whose generic callable signature makes `React.ComponentPropsWithRef<typeof X>` resolve to `{}`.

```tsx
import styled from 'styled-components';
import { Button } from '@mantine/core';

const Styled = styled(Button)`color: red;`;

// before: TS2769, `variant` and `children` rejected
// after: accepted
const usage = <Styled variant="filled">hi</Styled>;
```

When a target's props resolve to an empty `{}`, the styled component now falls back to accepting arbitrary props instead of rejecting all of them, both at the JSX call site and via `.attrs()`. Components whose prop types can be read are unaffected and stay fully type-checked.

### Details

- The detection keys on the props having no known keys. It's distributed over unions, so a bare union of disjoint shapes (`{ a: string } | { b: string }`, whose `keyof` is `never`) stays strict rather than going permissive. All-optional shapes (`{ a?: string }`) also stay strict.
- The permissive fallback is `unknown`-typed, not `any`, so reading props back off the component (`React.ComponentProps<typeof Styled>`) doesn't disable type-checking in consumer code.
- The widening is confined to the call surface and the attrs input; component statics and `defaultProps` keep their real prop types internally.

### Known limitation

For an un-introspectable target there is no base type to substitute against, so `as` is accepted but the `as`-target's own props aren't individually type-checked (e.g. `href` on `as="a"` isn't anchor-validated). An introspectable target keeps full `as` typing. A genuinely propless component (`() => <div/>`) is indistinguishable from an un-introspectable one and also becomes permissive (both resolve to `{}`).

## 2. `as`-target autocomplete survives mid-typing

On a polymorphic styled component rendering a different element through `as` (for example `as="video"`), starting to type a new prop name could make the whole attribute suggestion list vanish, so the rendered element's props (`loop`, `controls`, and the rest) disappeared while you were typing them. They now keep autocompleting as you go.

### Details

The previous shape exposed several call overloads. A half-typed or unknown attribute matched none of them, overload resolution failed, and JSX completion dropped to nothing. A single conditional call signature always resolves, so completion stays live. The conditional preserves the prior behaviors: a real `as`-target substitutes its props; no `as`, or an `as` typed as the wrapped component's own non-target type (e.g. Next.js `Link`'s `as?: Url`), keeps base props with `as` optional and assignable; `forwardedAs` substitutes the forwarded target.

A single generic signature does a little more type work per call site than the overload set did. The cost is kept small (target test against `string | AnyComponent` rather than the literal-heavy `WebTarget`, and a shared alias for the no-target result); net is roughly +5% type instantiations across the type test suite, with check time effectively unchanged.

## Verification

Verified against real `@mantine/core@7`: `React.ComponentPropsWithRef<typeof Button>` resolves to `{}`, the wrapped styled component now accepts arbitrary props and `children` (at the call site and through `.attrs()`), and normally typed styled components still reject unknown or mistyped props. Autocomplete was confirmed via the editor language service for both the mid-typing `as`-target case and plain usage. Type tests covering the Mantine factory shape, the union edge cases, and the existing `as` / `forwardedAs` behaviors are included; the full runtime suite (924 tests) passes.

Closes #5756

### CODE REVIEW

File: packages/styled-components/src/types.ts [hunk 1]
```diff
@@ -277,6 +277,18 @@ export interface IStyledComponentBase<
...
+ * Some wrapped targets can't be statically introspected and their props
+ * collapse to `{}` -- most notably polymorphic-factory components (e.g. Mantine
+ * v7's `Button`, `Card`, `Menu.Item`), whose generic callable signature defeats
+ * `React.ComponentPropsWithRef`. A closed `{}` would reject every prop at the JSX
+ * call site, including `children`. Falling back to a permissive prop bag keeps
+ * these components usable; targets with introspectable props are unchanged.
+ */
+type WidenUntypedProps<Props extends BaseObject> = keyof Props extends never
+  ? Props & { [key: string]: any }
+  : Props;
```
[the remark this answers is not in the document]
Comment (quantizor) [reply packages/styled-components/src/types.ts]: Good catch on both halves. Verified empirically against real `@mantine/core@7`: with the permissive bag, `<S as="a" href={123} />` does not error, so the `as`-target props do degrade to the catch-all. That turns out to be unavoidable while staying permissive: accepting unknown base props requires an index signature, and that same signature types every key (including the target's), so `Substitute<base, anchorProps>` resolves `href` to the catch-all regardless of which overload is selected. Setting `as`/`forwardedAs` to `never` can't coexist with "accept arbitrary props" here.

On the `any` leak: you were right, and it pointed at a real over-reach. The widening was applied to all of `IStyledComponentBase`, so it also hit `IStyledStatics`/`defaultProps`, where an `any` index signature was silently masking an assignability mismatch in the native render path. I confined the widening to `PolymorphicComponent` (the JSX call surface) only, which let me switch the bag to `unknown` as you suggested without leaking into the statics. Net type-perf on the existing workload actually improved slightly (-1953 instantiations). Fixed in be7390b6.

File: packages/styled-components/src/test/types.tsx [hunk 2]
```diff
@@ -815,3 +815,62 @@ const ToggleWithAttrs = styled(ToggleBase).attrs({ isOpen: false })``;
...
+`;
+// Arbitrary props + children accepted (permissive fallback).
+<StyledPolyButton variant="filled" color="blue">
+  hi
+</StyledPolyButton>;
+<StyledPolyButton>just children</StyledPolyButton>;
+// `as` polymorphism still resolves the target's props.
+<StyledPolyButton as="a" href="/x">
+  link
+</StyledPolyButton>;
```
[the remark this answers is not in the document]
Comment (quantizor) [reply packages/styled-components/src/test/types.tsx]: Confirmed by testing `<S as="a" href={123} />` against real `@mantine/core@7`: it does not error, so the polymorphic overload's target typing is not preserved for an un-introspectable base, and a wrong-`href` negative assertion would fail. This is inherent rather than a regression to guard against: the permissive bag required to accept the unknown base props types every key (including the `as`-target's) as the catch-all. I corrected the comment to state this explicitly instead of implying full `as` typing, so the test documents the real contract. Fixed in be7390b6.

File: packages/styled-components/src/types.ts [hunk 3]
```diff
@@ -268,11 +268,30 @@ export interface PolymorphicComponent<
...
+ * call site, including `children`. Falling back to a permissive prop bag keeps
+ * these components usable; targets with introspectable props are unchanged.
+ *
+ * Applied only to the JSX call surface (`PolymorphicComponent`), never to the
+ * statics (`IStyledStatics`, `defaultProps`), so internal code keeps the real
+ * `Props` and the widening can't leak past the call site.
+ */
+type WidenUntypedProps<Props extends BaseObject> = keyof Props extends never
+  ? Props & { [key: string]: unknown }
+  : Props;
```
[the remark this answers is not in the document]
Comment (quantizor) [reply packages/styled-components/src/types.ts]: Confirmed and fixed in e16ee494. Reproduced it directly: a bare union target like `(props: { a: string } | { b: string }) => ...` has `keyof` of `never` (union keys intersect), so the non-distributive check pulled it into the permissive branch and a bogus prop stopped erroring.

The check now distributes the "no known keys" test over `Props` first, so it widens only when every union constituent is truly empty:

```ts
type WidenUntypedProps<Props extends BaseObject> = (
  Props extends unknown ? (keyof Props extends never ? true : false) : never
) extends true
  ? Props & { [key: string]: unknown }
  : Props;
```

Added a regression test asserting a disjoint-union-prop target still rejects unknown props. Thanks for catching this.
